routers/api/v1/repo/issue.go

```diff
@@ -562,20 +564,25 @@ func EditIssue(ctx *context.APIContext, form api.EditIssueOption) {
 			return
 		}
 	}
-
-	if err = models.UpdateIssueByAPI(issue); err != nil {
-		ctx.Error(http.StatusInternalServerError, "UpdateIssueByAPI", err)
-		return
-	}
 	if form.State != nil {
-		if err = issue_service.ChangeStatus(issue, ctx.User, api.StateClosed == api.StateType(*form.State)); err != nil {
-			if models.IsErrDependenciesLeft(err) {
-				ctx.Error(http.StatusPreconditionFailed, "DependenciesLeft", "cannot close this issue because it still has open dependencies")
-				return
-			}
-			ctx.Error(http.StatusInternalServerError, "ChangeStatus", err)
+		issue.IsClosed = (api.StateClosed == api.StateType(*form.State))
+	}
+	statusChangeComment, titleChanged, err := models.UpdateIssueByAPI(issue, ctx.User)
```

Comment:
Rather than returning a `titleChanged` value and risk a race condition (from a title read in two different moments), it's probably better to return the "old" version of the issue, as evaluated by `UpdateIssueByAPI()`. We must keep in mind that `begin transaction` does not lock anything except in SQLite, so the title can change between the reading (above), the reading inside `UpdateIssueByAPI()` and the actual moment of the update.

I know we do this all over the place, but at least let's do it once. 😆

Reply:
The title changed comment is created within the session. This is done within the transaction.